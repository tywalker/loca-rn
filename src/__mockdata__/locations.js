import React from 'react';
import * as api from '../services/normalize';

const locationsJSON = {
  "places":{
    "place_type":"county",
    "total":9,
    "pages":1,
    "page":1,
    "bbox":"44.87621424504764,44.981884057971016,45.12378575495236,45.018115942028984",
    "place":[
      {
        "place_id":"8fIn8DVTUb7r2JsG",
        "woeid":"2346877",
        "latitude":"43.192",
        "longitude":"44.179",
        "place_url":"\/Russia\/North+Ossetia-Alania+Republic",
        "place_type":"region",
        "place_type_id":"8",
        "timezone":"Europe\/Moscow",
        "_content":"North Ossetia-Alania Republic, RU, Russia",
        "woe_name":"North Ossetia-Alania Republic"
      },
    ]
  }
}

export const mockLocations = () => {
  let combinedArr = [];

  placesJSON.places.place.map( places => {
    combinedArr = combinedArr.concat(places);
  })

  let nPlaces = api.normalizePlaces(combinedArr);

  return nPlaces;
}
