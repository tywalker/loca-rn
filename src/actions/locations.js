import {
  fetchPlacesFromBB,
  fetchPlaceChildren,
  fetchImages
} from '../services/api';

export const LAT_LON = "LAT_LON";
export const LOCATIONS_REQUEST = "LOCATIONS_REQUEST";
export const LOCATIONS_FAILURE = "LOCATIONS_FAILURE";
export const LOCATIONS_SUCCESS = "LOCATIONS_SUCCESS";
export const IMAGES_REQUEST = "IMAGES_REQUEST";
export const IMAGES_FAILURE = "IMAGES_FAILURE";
export const IMAGES_SUCCESS = "IMAGES_SUCCESS";

import { normalizePlaces, normalizeImages } from '../services/normalize';

export const latLon = (lat, lon) => {
  return {
    type: LAT_LON,
    lat,
    lon
  }
}
export const locationsSuccess = places => {
  return {
    type: LOCATIONS_SUCCESS,
    places
  }
};

export const locationsFailure = error => {
  return {
    type: LOCATIONS_FAILURE,
    error
  }
};

export const imagesSuccess = images => {
  return {
    type: IMAGES_SUCCESS,
    images
  }
}

export const imagesFailure = error => {
  return {
    type: IMAGES_FAILURE,
    error
  }
}

/**
 * Thunks
 */
export const locationsRequest = (bbox, distance) => {
  return function(dispatch) {
    fetchPlacesFromBB(bbox, distance)
      .then( res => res.data )
      .then( res => fetchPlaceChildren(res) )
      .then( res => {
        let combinedArr = [];

        res.map( places => {
          combinedArr = combinedArr.concat(places);
        })

        let nPlaces = normalizePlaces(combinedArr);

        dispatch(locationsSuccess(nPlaces));
      })
      .catch( error => dispatch(locationsFailure(error)) );
  }
};

export const imagesRequest = (places) => {
  return function(dispatch) {
    places.map( place => {
      fetchImages(place.id)
        .then( res => res.data )
        .then( res => {
          let resArr = res.photos.photo;
          let nImages = normalizeImages(resArr);

          dispatch(imagesSuccess(nImages));
        })
        .catch( error => dispatch(imagesFailure(error)) );

    })
  }
};
