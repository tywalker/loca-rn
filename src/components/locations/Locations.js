import React, { Component } from 'react';
import { FlatList, Image, StyleSheet, Text, View } from 'react-native';
import { connect } from 'react-redux';
import { getBoundingBoxFromGeo } from '../../services/api';

import { locationsRequest } from '../../actions/locations';

class Locations extends Component {

  componentDidUpdate(prevProps) {
    const { dispatch, lat, lon, places } = this.props;

    if (prevProps.lat !== lat && prevProps.lon !== lon) {
      let bbox = getBoundingBoxFromGeo(lat, lon, 5.0);

      dispatch(locationsRequest(bbox, 5.0));
    }
  }

  _keyExtractor = ( item, index ) => item.id;

  _renderItem = ({ item, index }) => (
    <View
      id={ index }
      style={ styles.location }
    >
      <Text style={ styles.locaLabel }>{ item.woe.name }</Text>
      <View style={ styles.locaImageContainer }>
        <Image
          source={{ uri: "https://placeimg.com/200/150/any" }}
          style={{ height: 150, width: 200 }}
        />
      </View>
    </View>
  );


  render() {
    const { places } = this.props;

    return (
      <View style={styles.container}>
        <View style={ styles.topNavigation }></View>
        <View style={ styles.mainViewContainer }>

          <FlatList
            data={ places }
            keyExtractor={ this._keyExtractor }
            renderItem={ this._renderItem }
          />

        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    backgroundColor: '#F5FCFF',
  },
  topNavigation: {
    flex: 0.2,
    backgroundColor: '#f0f0f0'
  },
  mainViewContainer: {
    flex: 0.8,
    backgroundColor: '#f8f8f8',
    alignItems: 'center'
  },
  locationsContainer: {
    flex: 1,
    minWidth: '90%',
    maxWidth: '90%',
    backgroundColor: '#f9f9f9'
  },
  location: {
    height: 300,
  },
  locaLabel: {
    color: "#555",
    fontSize: 16
  },
  locaImageContainer: {
    minHeight: 150,
    minWidth: 200
  }
});

function mapStateToProps(state) {
  return {
    places: state.locations.places,
    lat: state.locations.lat,
    lon: state.locations.lon
  }
}

export default connect(mapStateToProps)(Locations);
