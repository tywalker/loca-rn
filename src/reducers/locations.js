import { combinereducers } from 'redux';
import { contact_search_success } from '../actions';

const initialviews = { view: 'loading', scrollable: true };

function views( state = initialviews, action ) {
  switch (action.type) {
    case set_height_width:
      return {
        ...state,
        height: action.height,
        width: action.width
      };

    default:
      return state
  }
}

const locations = (state = {
  isfetching: false,
  didinvalidate: false,
  items: []
}, action) => {
  switch (action.type) {
    case request_posts:
      return {
        ...state,
        isfetching: true,
        didinvalidate: false
      }
    case receive_posts:
      return {
        ...state,
        isfetching: false,
        didinvalidate: false,
        items: action.posts,
        lastupdated: action.receivedat
      }
    default:
      return state
  }
};

const rootreducer = combinereducers({
  views,
  locations,
});

export default rootreducer
