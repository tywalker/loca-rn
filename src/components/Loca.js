import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';

import Locations from './locations/Locations';

class Loca extends Component {
  render() {
    return (
      <View style={ styles.container }>
        <Locations />
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    backgroundColor: '#F5FCFF',
  },
});

export default Loca;
