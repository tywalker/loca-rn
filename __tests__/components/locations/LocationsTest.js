import React from 'react';
import { FlatList, View } from 'react-native';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../../../src/reducers/locations';

import Loca from '../../../src/components/Loca';
import Locations from '../../../src/components/locations/Locations';
import { mockLocations } from '../../../src/mockdata/locations';

const mockStore = createStore(
  rootReducer,
  applyMiddleware(thunk),
);

let mockLocas = mockLocations();

const props = {
  locations: mockLocas,
  imagesDone: false,
};

describe("App loaded", () => {
  const wrapper = shallow(
    <Provider store={ mockStore }>
      <Locations store={ mockStore }/>
    </Provider>
  );

  it('renders Loca component', () => {
    expect(wrapper).toBeTruthy();
  });

  it('defaults view state to loading', () => {
    const state = mockStore.getState();

    expect(state.views.view).toBe('loading');
  });

  it('should successfully render the Locations component view', () => {
    wrapper.setProps({ imagesDone: false });
    let locations =  shallow(
      <Locations store={ mockStore }><View></View></Locations>);

    expect(locations.dive().dive()).toMatchSnapshot();
    wrapper.setProps({ imagesDone: true });
    expect(locations.dive().dive()).toMatchSnapshot();
  });

  it('should successfully update locations', () => {
    wrapper.setProps({ imagesDone: true, locations: mockLocas });

    let locations =  shallow(
      <Locations store={ mockStore } ><View></View></Locations>
    );

    console.log(props.locations)
    console.log(mockStore.getState())
    console.log(locations.dive().dive().debug());

    expect("false").toBe("false");
  });
});
