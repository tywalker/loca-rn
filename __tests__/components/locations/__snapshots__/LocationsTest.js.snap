// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App loaded should successfully render the Locations component view 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Component
    style={
      Object {
        "backgroundColor": "#F5FCFF",
        "flex": 1,
        "flexDirection": "column",
      }
    }
  >
    <Component
      style={
        Object {
          "backgroundColor": "#f0f0f0",
          "flex": 0.2,
        }
      }
    />
    <Component
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#f8f8f8",
          "flex": 0.8,
        }
      }
    >
      <Component />
    </Component>
  </Component>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Component
          style={
            Object {
              "backgroundColor": "#f0f0f0",
              "flex": 0.2,
            }
          }
        />,
        <Component
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#f8f8f8",
              "flex": 0.8,
            }
          }
        >
          <Component />
        </Component>,
      ],
      "style": Object {
        "backgroundColor": "#F5FCFF",
        "flex": 1,
        "flexDirection": "column",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "style": Object {
            "backgroundColor": "#f0f0f0",
            "flex": 0.2,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Component />,
          "style": Object {
            "alignItems": "center",
            "backgroundColor": "#f8f8f8",
            "flex": 0.8,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": "View",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Component
            style={
              Object {
                "backgroundColor": "#f0f0f0",
                "flex": 0.2,
              }
            }
          />,
          <Component
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#f8f8f8",
                "flex": 0.8,
              }
            }
          >
            <Component />
          </Component>,
        ],
        "style": Object {
          "backgroundColor": "#F5FCFF",
          "flex": 1,
          "flexDirection": "column",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "style": Object {
              "backgroundColor": "#f0f0f0",
              "flex": 0.2,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Component />,
            "style": Object {
              "alignItems": "center",
              "backgroundColor": "#f8f8f8",
              "flex": 0.8,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": "View",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`App loaded should successfully render the Locations component view 2`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Component
    style={
      Object {
        "backgroundColor": "#F5FCFF",
        "flex": 1,
        "flexDirection": "column",
      }
    }
  >
    <Component
      style={
        Object {
          "backgroundColor": "#f0f0f0",
          "flex": 0.2,
        }
      }
    />
    <Component
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#f8f8f8",
          "flex": 0.8,
        }
      }
    >
      <Component />
    </Component>
  </Component>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Component
          style={
            Object {
              "backgroundColor": "#f0f0f0",
              "flex": 0.2,
            }
          }
        />,
        <Component
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#f8f8f8",
              "flex": 0.8,
            }
          }
        >
          <Component />
        </Component>,
      ],
      "style": Object {
        "backgroundColor": "#F5FCFF",
        "flex": 1,
        "flexDirection": "column",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "style": Object {
            "backgroundColor": "#f0f0f0",
            "flex": 0.2,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Component />,
          "style": Object {
            "alignItems": "center",
            "backgroundColor": "#f8f8f8",
            "flex": 0.8,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": "View",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Component
            style={
              Object {
                "backgroundColor": "#f0f0f0",
                "flex": 0.2,
              }
            }
          />,
          <Component
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#f8f8f8",
                "flex": 0.8,
              }
            }
          >
            <Component />
          </Component>,
        ],
        "style": Object {
          "backgroundColor": "#F5FCFF",
          "flex": 1,
          "flexDirection": "column",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "style": Object {
              "backgroundColor": "#f0f0f0",
              "flex": 0.2,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Component />,
            "style": Object {
              "alignItems": "center",
              "backgroundColor": "#f8f8f8",
              "flex": 0.8,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": "View",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
